### PR TITLE
Add reason for price override

### PR DESCRIPTION
**Pull request by lkostrowski**
Dashboard: https://github.com/saleor/saleor-dashboard/pull/6732
Dummy-payment-app: https://github.com/saleor/apps/pull/2393
Port to main: once this is approved
Docs: https://github.com/saleor/saleor-docs/pull/1809

I want to merge this change because apps that set a custom price on a checkout line (via `HANDLE_CHECKOUTS`) currently leave no trace of **why** the price was overridden. This adds an optional `priceOverrideReason` the app can attach to the override, stored natively, carried onto the order line at completion, and surfaced on the dashboard for debugging/auditing.

## What changed

**Write — checkout only, app-only (`HANDLE_CHECKOUTS`)**
- New optional `priceOverrideReason` on `CheckoutLineInput` and `CheckoutLineUpdateInput` (`checkoutCreate` / `checkoutLinesAdd` / `checkoutLinesUpdate`).
- **Per-operation semantics** — the reason is tied to the price-setting operation: setting a `price` without a reason clears any previous reason; clearing the `price` clears the reason; a reason-only update is allowed when the line already has an override.
- **Coupling validation** — a reason with no incoming `price` and no existing override is rejected with the new `CheckoutErrorCode.PRICE_OVERRIDE_REASON_WITHOUT_OVERRIDE`.
- Blank/whitespace reasons normalize to `null`.
- `check_permissions_for_custom_prices` extended so setting a reason is app-only, exactly like `price`.

**Storage & conversion**
- New nullable `price_override_reason` (`CharField(255)`) on `CheckoutLine` and `OrderLine`.
- The reason is copied from the checkout line onto the `OrderLine` in `_create_line_for_order` during checkout completion.

**Read**
- `priceOverrideReason` exposed on the `CheckoutLine` type (`MANAGE_CHECKOUTS` / `HANDLE_CHECKOUTS`) and the `OrderLine` type (`MANAGE_ORDERS`). Subscription webhooks inherit it through these types; deprecated static payloads are untouched.

**Tests**
- Dedicated suite (`test_price_override_reason.py`, 17 tests) covering write, validation, normalization, per-operation update semantics, permission denial, checkout→order conversion, and read gating. Added test-first (red) then implemented (green).

# Impact

- [x] New migrations
- [x] New/Updated API fields or mutations
- [ ] Deprecated API fields or mutations
- [ ] Removed API types, fields, or mutations

# Docs

# Screens
CleanShot 2026-07-14 at 13 55 17@2x

CleanShot 2026-07-14 at 13 56 03@2x
CleanShot 2026-07-14 at 13 56 06@2x

### CODE REVIEW

File: saleor/graphql/checkout/mutations/utils.py [hunk 1]
```diff
@@ -429,6 +457,49 @@ def check_permissions_for_custom_prices(app, lines):
...
+        for line_info in (existing_lines_info or [])
+    }
+    for line_data in checkout_lines_data:
+        reason_is_set = (
+            line_data.custom_price_reason_to_update
+            and line_data.custom_price_reason is not None
+        )
+        if not reason_is_set:
+            continue
+        incoming_override = (
```
Comment (wcislo-saleor) [inline saleor/graphql/checkout/mutations/utils.py]: I think this validation is incomplete. Here's the test case showing an edge case.

```
def test_lines_update_clear_price_with_reason_raises(
    app_api_client, checkout_with_item, permission_handle_checkouts
):
    # given
    checkout = checkout_with_item
    line = checkout.lines.first()
    line.price_override = Decimal("10.00")
    line.price_override_reason = "old"
    line.save(update_fields=["price_override", "price_override_reason"])
    variant_id = graphene.Node.to_global_id("ProductVariant", line.variant.pk)

    # clearing the override and setting a reason in the same operation must be
    # rejected - the reason would otherwise be orphaned (no price override left).
    variables = {
        "id": to_global_id_or_none(checkout),
        "lines": [
            {
                "variantId": variant_id,
                "quantity": 1,
                "price": None,
                "priceOverrideReason": "still here",
            }
        ],
    }

    # when
    response = app_api_client.post_graphql(
        CHECKOUT_LINES_UPDATE, variables, permissions=[permission_handle_checkouts]
    )

    # then
    content = get_graphql_content(response)
    errors = content["data"]["checkoutLinesUpdate"]["errors"]
    assert len(errors) == 1
    assert (
        errors[0]["code"]
        == CheckoutErrorCode.PRICE_OVERRIDE_REASON_WITHOUT_OVERRIDE.name
    )
    assert (
        errors[0]["message"]
        == "Cannot set priceOverrideReason without a price override on the line."
    )
    line.refresh_from_db()
    assert line.price_override == Decimal("10.00")
    assert line.price_override_reason == "old"

```

File: saleor/graphql/checkout/mutations/utils.py [hunk 2]
```diff
@@ -429,6 +457,49 @@ def check_permissions_for_custom_prices(app, lines):
         )
 
 
+def _clean_price_override_reason(reason: str | None) -> str | None:
```
Comment (wcislo-saleor) [inline saleor/graphql/checkout/mutations/utils.py:474]: Strip reason to `max_length` of a field where it gets saved? Otherwise we're going to end up with a crash.

File: saleor/graphql/checkout/tests/mutations/test_price_override_reason.py
Comment (wcislo-saleor) [inline saleor/graphql/checkout/tests/mutations/test_price_override_reason.py:1]: `CLAUDE.md` says "When asserting GraphQL errors, assert error message too". Let's do that or remove it from `CLAUDE.md` if it doesn't make sense.

**Review by wcislo-saleor — CHANGES_REQUESTED** [review]

File: saleor/graphql/checkout/tests/mutations/test_price_override_reason.py
Comment (lkostrowski) [reply saleor/graphql/checkout/tests/mutations/test_price_override_reason.py:1]: It makes sense, just claude is not always respecting that, I dont know why :/ Maybe we need to move tests to skills, I will look into that separately

File: saleor/graphql/schema.graphql [hunk 3]
```diff
@@ -30934,6 +30953,13 @@ input CheckoutLineInput @doc(category: "Checkout") {
   """
   price: PositiveDecimal
 
+  """
+  Reason explaining why a custom `price` was set on the line, for debugging and auditing. Can be set only by apps with `HANDLE_CHECKOUTS` permission and only together with a `price` override. Blank values are stored as no reason. Limited to 255 characters; longer values are truncated.
```
Comment (wcislo-saleor) [inline saleor/graphql/schema.graphql]: Is there are a reason for this subtle difference in description between this `priceOverrideReason` and the other one?

... and only together with a `price` override
and
... and only when the line has a `price` override

The latter seems more accurate.

Comment (lkostrowski) [reply saleor/graphql/schema.graphql]: https://github.com/saleor/saleor/pull/19466/changes/dda7872acd355370c8c180e3a8b7c72312468cd3

**Review by wcislo-saleor — APPROVED** [review]
